integrations/gitlab/gitlab_integration/gitlab_service.py

```diff
@@ -683,6 +695,103 @@ async def get_all_issues(self, group: Group) -> typing.AsyncIterator[List[Issue]
+        self,
+        obj: RESTObject,
+        include_inherited_members: bool = False,
+        include_bot_members: bool = True,
+        include_public_email: bool = False,
+    ) -> dict[str, Any]:
+        """
+        Enriches an object (e.g., Project or Group) with its members.
+        """
+        members_list = []
+        async for members in self.get_all_object_members(
+            obj, include_inherited_members, include_bot_members
+        ):
+            if include_public_email:
+                tasks = [
+                    self.enrich_member_with_public_email(member) for member in members
+                ]
+                members_list.extend(await asyncio.gather(*tasks))
+            else:
+                members_list.extend(member.asdict() for member in members)
+
+        obj_dict: dict[str, Any] = obj.asdict()
+        obj_dict["__members"] = members_list
+        return obj_dict
+
+    async def get_all_object_members(
+        self,
+        obj: RESTObject,
+        include_inherited_members: bool = False,
+        include_bot_members: bool = True,
+    ) -> AsyncIterator[RESTObjectList]:
+        """
+        Fetches all members of an object (e.g., Project or Group) generically.
+        """
+        try:
+            obj_name = getattr(obj, "name", "unknown")
+            logger.info(f"Fetching all members of {obj_name}")
+
+            members_attr = "members_all" if include_inherited_members else "members"
+            members_manager = getattr(obj, members_attr, None)
+            if not members_manager:
+                raise AttributeError(f"Object does not have attribute '{members_attr}'")
+
+            fetch_func = members_manager.list
+
+            validation_func = functools.partial(
+                self.should_run_for_members, include_bot_members
+            )
+
+            async for members_batch in AsyncFetcher.fetch_batch(
+                fetch_func=fetch_func,
+                validation_func=validation_func,
+                pagination="offset",
+                order_by="id",
+                sort="asc",
+            ):
+                members: RESTObjectList = typing.cast(RESTObjectList, members_batch)
+
+                logger.info(
+                    f"Queried {len(members)} members {[member.username for member in members]} from {obj_name}"
+                )
+                yield members
+        except Exception as e:
+            logger.error(f"Failed to get members for object='{obj_name}'. Error: {e}")
+            return
+
+    async def enrich_member_with_public_email(
+        self, member: RESTObject
+    ) -> dict[str, Any]:
+        user: User = await self.get_user(member.id)
+        member_dict: dict[str, Any] = member.asdict()
+        member_dict["__public_email"] = user.public_email
+        return member_dict
+
+    async def get_user(self, user_id: str) -> User:
+        async with semaphore:
+            logger.info(f"fetching user {user_id}")
```

Comment:
is there more information that can be added to a `member` other than public email ?

Reply:
Here is a sample

```json
{
  "id": 15339857,
  "username": "tompo",
  "name": "Tom Kofi Tankilevitch",
  "state": "active",
  "locked": false,
  "avatar_url": "https://secure.gravatar.com/avatar/81e644d6312fa802fa1a11d605104ba745732a7054be8d3579d3024ff6168bcc?s=80&d=identicon",
  "web_url": "https://gitlab.com/tompo",
  "created_at": "2023-08-27T13:39:29.744Z",
  "bio": "",
  "location": "",
  "public_email": null,
  "skype": "",
  "linkedin": "",
  "twitter": "",
  "discord": "",
  "website_url": "",
  "organization": "",
  "job_title": "",
  "pronouns": null,
  "bot": false,
  "work_information": null,
  "followers": 0,
  "following": 0,
  "is_followed": false,
  "local_time": null
}

```

Reply:
Member

```json
{
  "id": 15339857,
  "username": "tempo",
  "name": "Tom Kofi Tankilevitch",
  "state": "active",
  "locked": false,
  "avatar_url": "https://secure.gravatar.com/avatar/81e644d6312fa802fa1a11d605104ba745732a7054be8d3579d3024ff6168bcc?s=80&d=identicon",
  "web_url": "https://gitlab.com/tom15339857",
  "access_level": 50,
  "created_at": "2023-08-27T13:39:29.744Z",
  "created_by": {
    "id": 13921399,
    "username": "matanlevi",
    "name": "Matan Levi",
    "state": "active",
    "locked": false,
    "avatar_url": "https://secure.gravatar.com/avatar/814e5edbb657a70c5c9bd09a4171768c245d18cabec5f3b440d43dbc26d36f88?s=80&d=identicon",
    "web_url": "https://gitlab.com/matanlevi"
  },
  "expires_at": null,
  "email": null,
  "membership_state": "active"
}

```